### PR TITLE
http keep alive settings investigation blog post

### PR DESCRIPTION
**Pull request by jaslakson**
Documenting our adventures with http requests in our Node code.

### CODE REVIEW

File: _posts/2017-01-17-http-keep-alive.markdown [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+// retry once for timeouts
+if (statusCode === 503) {
+  console.error('RETRY', url);
+  response = await fetch(url, fetchOptions).catch(catchError);
+  statusCode = response.status;
+}
+```
+
+With this new code, we were now able to pool our sockets which fixed the timeout issues and we were able to retry our requests in the case of a temporary blip. In the end we wish that we had addressed these issues prior to launch but the troubleshooting process was a great learning experience. In my years as a software engineer I have learned that no matter diligent you are in your process and review you can't catch everything. The key is to learn from your mistakes and improve for next time.
```
Comment (ebressler) [inline _posts/2017-01-17-http-keep-alive.markdown]: I think you missed a word.

`In my years as a software engineer I have learned that no matter` **how** `diligent you are in your process and review you can't catch everything.`

**Comment by ebressler** [comment]
I like it, however, your preformatted sections require some scrolling to see all the data, it might be worth breaking that into multiple lines or something.

File: _posts/2017-01-17-http-keep-alive.markdown [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+At this point we were also seeing an H13 error in our Heroku logs. From Heroku's error documentation "This error is thrown when a process in your web dyno accepts a connection, but then closes the socket without writing anything to it."
+
+Example H13 error
+```
+2010-10-06T21:51:37-07:00 heroku[router]: at=error code=H13 desc="Connection closed without response" method=GET path="/" host=myapp.herokuapp.com fwd=17.17.17.17 dyno=web.1 connect=3030ms service=9767ms status=503 bytes=0
+```
+
+After trying a quick-fix we realized that while we searched for a fix we had to, at the very least, provide slightly better error page for our users. The 500 errors that they were seeing were being displayed after the max-timeout was surpassed - which was 30 seconds! Additionally, the error page was the standard Heroku error page which does not inspire confidence. So, we set out to implement a custom timeout response and install some extra logging to help us gather some additional data.
```
Comment (aczerepinski) [inline _posts/2017-01-17-http-keep-alive.markdown]: provide `a` slightly

File: _posts/2017-01-17-http-keep-alive.markdown [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+To do this, needed to inject some custom middleware in our Express pipeline. To start, we set the timeout at 25s, 5s short of the Heroku timeout.
```
Comment (aczerepinski) [inline _posts/2017-01-17-http-keep-alive.markdown]: this, `we` needed

**Comment by jaslakson** [comment]
@ebressler and @aczerepinski - Thanks for the review. I'll spruce it up a bit and add all the missing words!

**Pull request merged**